enterprise/coderd/prebuilds/metricscollector.go

```diff
@@ -0,0 +1,126 @@
+package prebuilds
+
+import (
+	"context"
+	"time"
+
+	"cdr.dev/slog"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+	"github.com/coder/coder/v2/coderd/prebuilds"
+)
+
+var (
+	labels               = []string{"template_name", "preset_name", "organization_name"}
+	createdPrebuildsDesc = prometheus.NewDesc(
+		"coderd_prebuilds_created_total",
+		"The number of prebuilds that have been created to meet the desired count set by presets.",
+		labels,
+		nil,
+	)
+	failedPrebuildsDesc = prometheus.NewDesc(
+		"coderd_prebuilds_failed_total",
+		"The number of prebuilds that failed to build during creation.",
+		labels,
+		nil,
+	)
+	claimedPrebuildsDesc = prometheus.NewDesc(
+		"coderd_prebuilds_claimed_total",
+		"The number of prebuilds that were claimed by a user. Each count means that a user created a workspace using a preset and was assigned a prebuild instead of a brand new workspace.",
+		labels,
+		nil,
+	)
+	usedPresetsDesc = prometheus.NewDesc(
+		"coderd_prebuilds_used_presets",
+		"The number of times a preset was used to build a prebuild.",
+		labels,
+		nil,
+	)
+	desiredPrebuildsDesc = prometheus.NewDesc(
+		"coderd_prebuilds_desired",
+		"The number of prebuilds desired by each preset of each template.",
+		labels,
+		nil,
+	)
+	runningPrebuildsDesc = prometheus.NewDesc(
+		"coderd_prebuilds_running",
+		"The number of prebuilds that are currently running. Running prebuilds have successfully started, but they may not be ready to be claimed by a user yet.",
+		labels,
+		nil,
+	)
+	eligiblePrebuildsDesc = prometheus.NewDesc(
+		"coderd_prebuilds_eligible",
+		"The number of eligible prebuilds. Eligible prebuilds are prebuilds that are ready to be claimed by a user.",
+		labels,
+		nil,
+	)
+)
+
+type MetricsCollector struct {
+	database    database.Store
+	logger      slog.Logger
+	snapshotter prebuilds.StateSnapshotter
+}
+
+var _ prometheus.Collector = new(MetricsCollector)
+
+func NewMetricsCollector(db database.Store, logger slog.Logger, snapshotter prebuilds.StateSnapshotter) *MetricsCollector {
+	return &MetricsCollector{
+		database:    db,
+		logger:      logger.Named("prebuilds_metrics_collector"),
+		snapshotter: snapshotter,
+	}
+}
+
+func (*MetricsCollector) Describe(descCh chan<- *prometheus.Desc) {
+	descCh <- createdPrebuildsDesc
+	descCh <- failedPrebuildsDesc
+	descCh <- claimedPrebuildsDesc
+	descCh <- usedPresetsDesc
+	descCh <- desiredPrebuildsDesc
+	descCh <- runningPrebuildsDesc
+	descCh <- eligiblePrebuildsDesc
+}
+
+func (mc *MetricsCollector) Collect(metricsCh chan<- prometheus.Metric) {
+	ctx, cancel := context.WithTimeout(dbauthz.AsPrebuildsOrchestrator(context.Background()), 10*time.Second)
+	defer cancel()
+	// nolint:gocritic // just until we get back to this
```

Comment:
Is this a TODO? Is there a ticket for it? Implications?

Reply:
Moved the comment to above line 82 where it belongs; we need to set an auth context, this one is the most correct.